### PR TITLE
Add dag support

### PR DESCRIPTION
**Pull request by schrodit**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
`artifactsFrom` are currently not implemented and will come in a different PR.

**Release note**:

```improvement operator
Refactor old testflow to use DAG style workflows
```
```action user
:warning: Structure of old testflow are no longer working. Please adjust to the new dag style.
```
```action user
:warning: `testFlow` is renamed to `testflow`.
```

### CODE REVIEW

File: examples/local-testrun.yaml [hunk 1]
```diff
@@ -40,10 +40,14 @@ spec:
 
   # the execution flow:
   testFlow:
```
[the remark this answers is not in the document]
Comment (schrodit) [reply examples/local-testrun.yaml]: its `testflow`

File: pkg/testmachinery/testflow/flow.go [hunk 2]
```diff
@@ -124,82 +83,57 @@ func (f *Flow) GetVolumes() []corev1.Volume {
...
 
-func (f *Flow) addConfigToTestDefinition(step *Step, td *testdefinition.TestDefinition) error {
-	cfg := config.New(step.Info.Config)
-	if err := td.AddConfig(cfg); err != nil {
-		return err
-	}
-	if err := td.AddConfig(f.globalConfig); err != nil {
-		return err
+// GetStatus returns the status of all nodes of the current flow.
+func (f *Flow) GetStatus() []*tmv1beta1.StepStatus {
```
Comment (OlegLoewen) [inline pkg/testmachinery/testflow/flow.go]:
```suggestion
func (f *Flow) GetStatuses() []*tmv1beta1.StepStatus {
```

File: pkg/testmachinery/testflow/flow_operations.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+			Nodes: nodes,
+		}
+	}
+	return stepMap, testdefinitions, usedLocations, nil
+}
+
+// CreateInitialDAG creates a DAG by evaluating the dependsOn steps.
+func CreateInitialDAG(steps map[string]*Step, root *node.Node) {
+	for _, step := range steps {
+		if step.Info.DependsOn == nil || len(step.Info.DependsOn) == 0 {
```
Comment (OlegLoewen) [inline pkg/testmachinery/testflow/flow_operations.go]:
```suggestion
		if step.Info.DependsOn == nil || len(step.Info.DependsOn) == 0 {
			addRootNodeAsParent(step)
		} else {
			addDependendsAsParent(step)
		}
```
Decouple a little more for better reading

File: pkg/testmachinery/testflow/node/node.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+	}
+	if td.Location != nil {
+		status.TestDefinition.Location = *td.Location.GetLocation()
+	}
+
+	return status
+}
+
+// SetOutput adds std output to the test and marks the node as node with output.
+func (n *Node) SetOutput() {
```
Comment (OlegLoewen) [inline pkg/testmachinery/testflow/node/node.go]:
```suggestion
func (n *Node) EnableOutput() {
```
For better reading

File: pkg/testmachinery/testflow/validation.go [hunk 5]
```diff
@@ -16,59 +16,85 @@ package testflow
...
+		if err := ValidateDefinition(identifier, step.Definition); err != nil {
+			return err
+		}
 
-			for elemIndex, elem := range step.Config {
-				if err := config.Validate(fmt.Sprintf("%s.config.[%d]", identifier, elemIndex), elem); err != nil {
-					return err
-				}
-			}
+		testdefinitions, err := locs.GetTestDefinitions(step.Definition)
```
Comment (OlegLoewen) [inline pkg/testmachinery/testflow/validation.go]:
```suggestion
		testDefinitions, err := locs.GetTestDefinitions(step.Definition)
```

File: pkg/testmachinery/testflow/validation.go [hunk 6]
```diff
@@ -16,59 +16,85 @@ package testflow
...
 
-			for _, td := range testdefinitions {
-				if err := testdefinition.Validate(fmt.Sprintf("%s; Location: \"%s\"; File: \"%s\"", identifier, td.Location.Name(), td.FileName), td.Info); err != nil {
-					return err
+		usedStepNames[step.Name] = true
+		usedTestdefinitions += len(testdefinitions)
+	}
+
+	// check if dependent steps exist
+	for i, step := range tf {
```
Comment (OlegLoewen) [inline pkg/testmachinery/testflow/validation.go]:
```suggestion
	err := checkIfDependentStepsExist(tf)
```
decouple

File: pkg/controller/testrun_control_update.go [hunk 7]
```diff
@@ -97,56 +93,62 @@ func (r *TestrunReconciler) completeTestrun(tr *tmv1beta1.Testrun, wf *argov1.Wo
...
 	tr.Status.State = fmt.Sprintf("Testmachinery executed %d/%d Steps", completedSteps, numSteps)
 }
 
+func getNodeStatusByName(wf *argov1.Workflow, templateName string) *argov1.NodeStatus {
+	for _, node := range wf.Status.Nodes {
+		if node.TemplateName == templateName {
+			return &node
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (schrodit) [reply pkg/controller/testrun_control_update.go]: not necessary.
When a testrun is created but the workflow is still initializing or not all dag nodes are fully calculated.

File: pkg/testmachinery/testflow/flow_operations.go (hunk 3 above)
Comment (schrodit) [reply pkg/testmachinery/testflow/flow_operations.go]: done for the `addDependentStepAsParent` function.
I think the first decoupling is unecessary as there are only 2 commands?

File: pkg/controller/testrun_control_update.go [hunk 8]
```diff
@@ -97,56 +93,62 @@ func (r *TestrunReconciler) completeTestrun(tr *tmv1beta1.Testrun, wf *argov1.Wo
 
 func updateStepsStatus(tr *tmv1beta1.Testrun, wf *argov1.Workflow) {
```
[the remark this answers is not in the document]
Comment (schrodit) [reply pkg/controller/testrun_control_update.go]: done

File: examples/local-testrun.yaml [hunk 9]
```diff
@@ -39,11 +43,16 @@ spec:
...
+    dependsOn: [ demo-step ]
+    definition:
+      label: default
 
   onExit:
-    - - name: DeleteShootTestDef
-        condition: error
+  - name: exit
+    definition:
+      name: exit-hander
```
[the remark this answers is not in the document]
Comment (schrodit) [reply examples/local-testrun.yaml]: done

File: pkg/testmachinery/testflow/validation.go (hunk 6 above)
Comment (schrodit) [reply pkg/testmachinery/testflow/validation.go]: done

File: pkg/testmachinery/testflow/validation.go (hunk 5 above)
Comment (schrodit) [reply pkg/testmachinery/testflow/validation.go]: done

File: pkg/testmachinery/testflow/node/node.go (hunk 4 above)
Comment (schrodit) [reply pkg/testmachinery/testflow/node/node.go]: done

File: pkg/testmachinery/testflow/flow_operations.go (hunk 3 above)
Comment (OlegLoewen) [reply pkg/testmachinery/testflow/flow_operations.go]: This is also fine. It's just, if the code inside a function is consistently on the same abstraction level, it's better to read.

**Pull request merged**